src/Behat/Behat/EventDispatcher/Event/GherkinEventFactory.php

```diff
@@ -0,0 +1,158 @@
+<?php
+
+/*
+ * This file is part of the Behat.
+ * (c) Konstantin Kudryashov <ever.zet@gmail.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Behat\Behat\EventDispatcher\Event;
+
+use Behat\Behat\Tester\Context\BackgroundContext;
+use Behat\Behat\Tester\Context\ScenarioContext;
+use Behat\Behat\Tester\Context\StepContext;
+use Behat\Behat\Tester\Result\StepResult;
+use Behat\Testwork\EventDispatcher\Event\EventFactory;
+use Behat\Testwork\EventDispatcher\Event\ExerciseEventFactory;
+use Behat\Testwork\Tester\Context\TestContext;
+use Behat\Testwork\Tester\Context\SpecificationContext;
+use Behat\Testwork\Tester\Result\TestResult as Result;
+use Behat\Testwork\Tester\Setup\Setup;
+use Behat\Testwork\Tester\Setup\Teardown;
+
+/**
+ * Gherkin-based event factory.
+ *
+ * @author Konstantin Kudryashov <ever.zet@gmail.com>
+ */
+final class GherkinEventFactory implements EventFactory
+{
+    /**
+     * @var ExerciseEventFactory
+     */
+    private $exerciseFactory;
+
+    /**
+     * Initializes event factory.
+     *
+     * @param ExerciseEventFactory $factory
+     */
+    public function __construct(ExerciseEventFactory $factory)
+    {
+        $this->exerciseFactory = $factory;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function createBeforeTestedEvent(TestContext $context)
+    {
+        switch (true) {
+            case $context instanceof StepContext:
+                return new BeforeStepTested($context);
+
+            case $context instanceof BackgroundContext:
+                return new BeforeBackgroundTested($context);
+
+            case ($context instanceof ScenarioContext && $context->isOutline()):
+                return new BeforeOutlineTested($context);
+
+            case ($context instanceof ScenarioContext && $context->isExample()):
+                return new BeforeExampleTested($context);
+
+            case $context instanceof ScenarioContext:
+                return new BeforeScenarioTested($context);
+
+            case $context instanceof SpecificationContext:
+                return new BeforeFeatureTested($context);
+        }
+
+        return $this->exerciseFactory->createBeforeTestedEvent($context);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function createAfterSetupEvent(TestContext $context, Setup $setup)
+    {
+        switch (true) {
```

Comment:
Switching on the type of the context suggests generation of the events could be responsibility of the context itself